xbmc/utils/GUIUtils.cpp

```diff
@@ -0,0 +1,363 @@
+{
+  if (itemToPlay->m_bIsFolder) // build a playlist and play it
+  {
+    XFILE::CMusicDatabaseDirectory dir;
+    // Don't add top level nodes to the playlist - should be excluded by the calling code already
+    XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE Node =
+        dir.GetDirectoryParentType(itemToPlay->GetPath());
+    if (Node == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_OVERVIEW)
+      return false;
+    CFileItemList queuedItems;
+    AddItemToPlayList(itemToPlay, queuedItems);
+    CServiceBroker::GetPlaylistPlayer().ClearPlaylist(PLAYLIST_MUSIC);
+    CServiceBroker::GetPlaylistPlayer().Reset();
+    CServiceBroker::GetPlaylistPlayer().Add(PLAYLIST_MUSIC, queuedItems);
+    CServiceBroker::GetPlaylistPlayer().SetCurrentPlaylist(PLAYLIST_MUSIC);
+    CServiceBroker::GetPlaylistPlayer().Play();
+  }
+  else // song, so just play it
+  {
+    CFileItem item(itemToPlay->GetItemToPlay());
+    PlaySong(item);
+  }
+  m_musicDatabase.Close();
+  return true;
+}
+
+bool CGUIUtils::AddItemToPlayList(const CFileItemPtr& pItem, CFileItemList& queuedItems)
+{
+  int jobid = CJobManager::GetInstance().AddJob(new CAddItemToPlaylistJob(pItem, queuedItems),
+                                                nullptr, CJob::PRIORITY_NORMAL);
+
+  if (!CGUIDialogBusy::WaitOnEvent(m_utilsJobEvent, 500)) // 500ms before busy dialog appears
+  {
+    // Cancel job if its still waiting in queue (unlikely) or still running
+    CJobManager::GetInstance().CancelJob(jobid);
+    return false;
+  }
+  return true;
+}
+
+void CGUIUtils::PlaySong(CFileItem& item)
+{
+  if (item.HasMusicInfoTag())
+    CServiceBroker::GetPlaylistPlayer().Play(std::make_shared<CFileItem>(item), "");
+}
+
+void CGUIUtils::AddToPlayList(const CFileItemPtr& pItem, CFileItemList& queuedItems)
+{
+  if (!pItem->CanQueue() || pItem->IsRAR() || pItem->IsZIP() || pItem->IsParentFolder())
+    return;
+
+  // fast lookup is needed here
+  queuedItems.SetFastLookup(true);
+
+  if (pItem->IsMusicDb() && pItem->m_bIsFolder && !pItem->IsParentFolder())
+  { // we have a music database folder, just grab the "all" item underneath it
+    XFILE::CMusicDatabaseDirectory dir;
+
+    if (!dir.ContainsSongs(pItem->GetPath()))
+    { // grab the ALL item in this category
+      // Genres will still require 2 lookups, and queuing the entire Genre folder
+      // will require 3 lookups (genre, artist, album)
+      CMusicDbUrl musicUrl;
+      if (musicUrl.FromString(pItem->GetPath()))
+      {
+        musicUrl.AppendPath("-1/");
+        CFileItemPtr item(new CFileItem(musicUrl.ToString(), true));
+        item->SetCanQueue(true); // workaround for CanQueue() check above
+        AddToPlayList(item, queuedItems);
+      }
+      return;
+    }
+  }
+  if (pItem->m_bIsFolder)
+  {
+    // Check if we add a locked share
+    if (pItem->m_bIsShareOrDrive)
+    {
+      CFileItem item = *pItem;
+      if (!g_passwordManager.IsItemUnlocked(&item, "music"))
```

Comment:
Why do you first make a copy of `pItem` which is a pointer, just to pass a pointer again in the next line? Is the called method manipulating the item passed in so you need a copy to not to change the instance behind `pItem`?